### PR TITLE
[DI] Introducing autoconfigure: automatic _instanceof configuration

### PR DESCRIPTION
**Pull request by weaverryan**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes (mostly, a continuation of a new feature)
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | n/a
| License       | MIT
| Doc PR        | https://github.com/symfony/symfony-docs/issues/7538

This is a proposal to allow the user to opt into some automatic `_instanceof` config. Suppose I want to auto-tag all of my voters and event subscribers

```yml
# current
services:
    _defaults:
        autowire: true

    _instanceof:
        Symfony\Component\Security\Core\Authorization\Voter\VoterInterface:
            tags: [security.voter]

        Symfony\Component\EventDispatcher\EventSubscriberInterface:
            tags: [kernel.event_subscriber]

    # services using the above tags
    AppBundle\Security\PostVoter: ~
    AppBundle\EventListener\CheckRequirementsSubscriber: ~
```

If I'm registering a service with a class that implements `VoterInterface`, when would I ever *not* want that to be tagged with `security.voter`? Here's the proposed code:

```yml
# proposed
services:
    _defaults:
        autowire: true
        autoconfigure: true

    # services using the auto_configure_instanceof functionality
    AppBundle\Security\PostVoter: ~
    AppBundle\EventListener\CheckRequirementsSubscriber: ~
```

The user must opt into this and it only applies locally to this configuration file. It works because each enabled bundle would have the opportunity to add one or more "automatic instanceof" definitions - e.g. SecurityBundle would add the `security.voter` instanceof config, FrameworkBundle would add the `kernel.event_subscriber` instanceof config, etc.

For another example, you can check out the proposed changes to `symfony-demo` - symfony/symfony-demo#483 - the `_instanceof` section is pretty heavy: https://github.com/hhamon/symfony-demo/blob/81694ac21e63bebe7ecfa22fa689766f2f38ccd5/app/config/services.yml#L20

Thanks!

### CODE REVIEW

**Comment by iltar** [comment]
Would it be possible for 3rd party bundles to extend this functionality? Say my Bundle has an interface that works with a tag as well, it would be very nice for developers to automatically have the tags.

**Comment by weaverryan** [comment]
@iltar absolutely - and for that exact reason - auto-tagging by interface for a tag introduced by a third-party bundle :)

**Comment by weaverryan** [comment]
This is ready! I added a note about an issue with the (already-merged) `instanceof` behavior and tags, which prevents us from adding a few *more* automatic tags. We can/should work that out over the next month.

I would *love* to sneak this in for 3.3 - the current `_instanceof` lacks the DX it's meant to help! :).

File: src/Symfony/Component/DependencyInjection/ContainerBuilder.php
Comment (dunglas) [inline src/Symfony/Component/DependencyInjection/ContainerBuilder.php]: Maybe should you add a description to those public methods?

File: src/Symfony/Component/DependencyInjection/Loader/YamlFileLoader.php
Comment (dunglas) [inline src/Symfony/Component/DependencyInjection/Loader/YamlFileLoader.php]: Nested ternaries are hard to read :) Why not using `if` blocks? In Symfony 4 we'll be able to replace such constructs with `??`.

File: src/Symfony/Component/DependencyInjection/Loader/YamlFileLoader.php
Comment (dunglas) [inline src/Symfony/Component/DependencyInjection/Loader/YamlFileLoader.php]: This whole block dans be simplified:

```php
if (isset($service['auto_configure_instanceof'])) {
    $definition->setShouldAutoConfigureInstanceofConditionals($service['auto_configure_instanceof']);
} elseif (isset($defaults['auto_configure_instanceof'])) {
    $definition->setShouldAutoConfigureInstanceofConditionals($defaults['auto_configure_instanceof']);
}
```

**Review by dunglas — CHANGES_REQUESTED** [review]

**Comment by dunglas** [comment]
I really like this proposal! Just some ideas to go further:

Why putting `auto_configure_instanceof` as a child of the `_defaults` key? It's not really a default. What do you think of:

```
services:
  auto_instanceof: true
  # ...
```

And I wonder if this approach should not totally replace the actual `_instanceof` key? In my experience, `_instanceof` sections in config files rapidly get ugly, hard to read and to debug.
In most cases, such should be done in bundles. Sometimes globally for an app. But I struggle to find a valid use case to do it in a specific config file. Doing it programmatically in bundles or in the app's kernel as you suggest looks more elegant to me. It's an advanced feature, and people will not have to deal often with it.
In the same time, removing the support for config files will ease the maintenance.

**Comment by dunglas** [comment]
Btw, ActionBundle only allows to configure the tag/interface mapping globally and having the ability to do it per file has never been requested.

**Comment by sstok** [comment]
👍 for 
```
services:
  _auto_instanceof: true
  # ...
```

Instead of the defaults.

About the replacing of the `_instanceof`, I'm not fully sure 🤔 we can't remove it now because we are already in feature free. And I'm actually using this 😅 but for the future it would make usage much more powerful 👍

**Comment by dunglas** [comment]
> we can't remove it now because we are already in feature freeze

The feature freeze is exactly for this kind of fine tuning :)

**Comment by sstok** [comment]
OK 🙂 https://imgflip.com/i/1mncy0

**Comment by nicolas-grekas** [comment]
Having the option in `_default` is relevant to me. It makes it conceptually aligned with the fact that this is a per-definition setting, which can be enabled on a file by file basis thanks to `_defaults`, but can also be *disabled* case by case.

About the name, what about just using `autoconfigure`?
```yaml
services:
  _defaults:
    autowire: true
    autoconfigure: true
```

File: src/Symfony/Component/DependencyInjection/Definition.php
Comment (GuilhemN) [inline src/Symfony/Component/DependencyInjection/Definition.php]: What about removing the `should`? It doesn't bring anything imo.

Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Definition.php]: of just "isAutoconfigured"

**Comment by weaverryan** [comment]
I will still update from your comments and finish this - I think this is an easy and big win that makes the new `instanceof` work well. But, I'm waiting for #22294, which will affect the code in this one a bunch :)

**Comment by fabpot** [comment]
:+1:

**Comment by weaverryan** [comment]
This is ready to go! The configuration is now called `instanceof` and I auto-tagged all tags that make sense (using `UnusedTagsPass` as a guide) - even if they are a rarely-used tag (so that we're consistent).

As we've mentioned in a few other places, this makes it more likely that a tag will be applied to a service twice, but in #22398 I illustrate how the tags are already added in a specific order so that compiler passes can handle (de-dup) this situation. In #22396 (and related changes that are needed to 3.2, once that's merged up), we're fixing individual compiler passes that may not handle multiple tags gracefully yet. Most already handle them perfectly (because multi tags was always a valid situation).

Cheers!

File: src/Symfony/Component/DependencyInjection/ContainerBuilder.php
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/ContainerBuilder.php]: could this return the ChildDefinition instead of this second arg?
that would allow things like `$container->addAutomaticInstanceofDefinition(\Twig_ExtensionInterface::class)->addTag('twig.extension')`

File: src/Symfony/Component/DependencyInjection/Loader/YamlFileLoader.php
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Loader/YamlFileLoader.php]: not sure it makes sense here? if it does, it's missing from the xsd

File: src/Symfony/Component/DependencyInjection/ContainerBuilder.php
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/ContainerBuilder.php]: about the name, `$container->registerForAutoconfiguration()`?

File: src/Symfony/Component/DependencyInjection/Loader/YamlFileLoader.php
Comment (weaverryan) [inline src/Symfony/Component/DependencyInjection/Loader/YamlFileLoader.php]: Ha, so now we've reversed roles :).

Maybe? I have an IntegrationTest testing the override logic for it, so that's solid. In theory, I could use `autoconfigure: true` under `_defaults`, but then set `autoconfigure: false` for one specific interface (e.g. `VoterInterface`) because I'm doing something weird and want to control all of my voters entirely myself.

File: src/Symfony/Component/DependencyInjection/ContainerBuilder.php
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/ContainerBuilder.php]: Append, fail or replace that's my only remaining questions

Comment (weaverryan) [inline src/Symfony/Component/DependencyInjection/ContainerBuilder.php]: Definitely appending (which means, merging). 

Why?

1) If 2 DI extensions both autoconfigure the same instance (which probably shouldn't happen), we should probably **throw an exception** or **merge** them together. Replacing seems completely wrong, since execution order isn't guaranteed. Of course, merging is a bit weird too, since execution order isn't guaranteed. But if they configure different things, merging is great. If they configure the same config... well, they shouldn't do that :).

2) If the end-user wants to auto configure an instanceof globally, why not let them? e.g. suppose they want to add some extra config to all VoterInterface? In this case, throwing an exception does *not* make sense. So, **replacing** or **merging** makes sense. And I think merging is more user-friendly: they're able to inherit all the original auto-config, then override whatever they want. They will get duplicate tags... but since new tags will be added first, they will override the original tags.

So, appending/merging makes sense to me. This actually affects 2 places in the code - calling `registerForAutoconfiguration()` 2 times on the same builder and merging 2 builders... but I think they should probably act the same.

**Comment by weaverryan** [comment]
Changes made!

**Comment by weaverryan** [comment]
Updated again: the PR description accurately explains the feature.

Most recent change is that:

A) If you call `ContainerBuilder::registerForAutoconfiguration()` multiple times, you receive the same `ChildDefinition` each time, allowing you to edit it directly. This is an edge-case, but would allow the end-user to have full control over the autoconfig for a specific instance (e.g. via a compiler pass)

B) In `ContainerBuilder::merge()`, if the same interface has been autoconfigured, we throw an exception. This is mainly to prevent 2 DI extensions from modifying the *same* interface... which is a problem because execution order isn't guaranteed. Configuring 2 instances is still totally possible, you just need to do it in a compiler pass (but probably shouldn't be done).

**Comment by weaverryan** [comment]
Test failures are unrelated - will be fixed by #22414

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php
Comment (stof) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: not a good idea: it has a mandatory attribute (the `extended_type` one), so auto_configuration would not help

**Review by stof — CHANGES_REQUESTED** [review]

**Comment by stof** [comment]
@weaverryan please check other tags to see whether they have required attributes. I don't remember it for all of them.

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php
Comment (weaverryan) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: Fixed! Thank you - I was trying to be careful not to select tags like that - obviously missed on. Unrelated, that required attribute is a bummer for the user since it's also in the class itself :)

Comment (stof) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: sure, but calling the method requires instantiation, which would break lazy-loading.

**Comment by weaverryan** [comment]
I just re-audited all the tags: there were a few others that needed to be removed (unfortunately) because they have required attributes. 

I also think there are more tags hiding that could be applied. I used `UnusedTagsPass` as my guide, but what about `property_info.list_extractor`, `property_info.type_extractor` or `cache.pool`? I think `UnusedTagsPass` doesn't have everything...

**Comment by nicolas-grekas** [comment]
I can answer for `cache.pool`: there is no autoconf to do for them. The tag is not linked to some behavior but to constructor wiring, so not "instanceof" related.

**Comment by weaverryan** [comment]
Sorry for the noise - I think I've got them all now.

This is ready to go!

**Comment by stof** [comment]
> I think UnusedTagsPass doesn't have everything...

@weaverryan I think this is totally possible. The pass might have been forgotten when adding new tags after it was introduced in Symfony.

**Review by nicolas-grekas — APPROVED** [review]
:+1:

**Comment by weaverryan** [comment]
This is ready again - tests passing `use` statements in `FrameworkExtension` re-ordered!

**Comment by weaverryan** [comment]
I just added one more autoconfigured tag: `controller.service_arguments` for `AbstractController`. If you're extending this, then your controller is registered as a service. And I think you should benefit from the argument autowiring automatically.

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php
Comment (aitboudad) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: wouldn't be better to check if form config is enabled before registering `FormTypeInterface` and `FormTypeGuesserInterface`? 
The same thing should applied for optional components like property_info, serializer, validator.

Comment (nicolas-grekas) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: That's already conditionally applied so IMHO that wouldn't provide much benefit to do so.

Comment (weaverryan) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: I don't think it makes any difference, but I can see the appeal of organizing these in that way (i.e. adding the serializer-related stuff inside `registerSerializerConfiguration`). We could always move it later without a BC break since adding these when the form is disabled (for example) has no effect.

**Comment by weaverryan** [comment]
After talking with Nicolas, I just added `controller.service_arguments` also to `Controller`. It's actually 100% valid to register your class as a service but extend `\Controller`. In fact, everything 100% works, the only difference is that those classes have access to `$this->container`.

**Comment by nicolas-grekas** [comment]
We could also add setters to the list, base on "Aware" interfaces:
- LoggerAwareInterface => `setLogger(new Reference('logger'))`
- ContainerAwareInterface  => `setContainer(new Reference('service_container'))`
- EventManagerAwareInterface
- RequestContextAwareInterface
- DenormalizerAwareInterface
- NormalizerAwareInterface
- SerializerAwareInterface
- InputAwareInterface
- ObjectManagerAware => in DoctrineBundle?

File: src/Symfony/Component/DependencyInjection/ContainerBuilder.php
Comment (fabpot) [inline src/Symfony/Component/DependencyInjection/ContainerBuilder.php]: missing `"` around `%s`

**Comment by fabpot** [comment]
Thank you @weaverryan.

**Comment by TomasVotruba** [comment]
Just wondering, is there a case you don't want this feature on by default?

This will only force people to put same lines in every config file they have.

**Comment by curry684** [comment]
> Just wondering, is there a case you don't want this feature on by default?

That's just not how Symfony works. If the feature were on by default it would be "magic", the killer word used to describe "something is happening in my application and I never told it to". Magic can actually be good. Magic can be a conscious choice. Laravel for example is built around magic, and is used by millions of developers madly in love with its magic. Magic makes sure you can develop quick and not have to know everything about what you're doing.

Many professional developers despise magic, because it allows people to have no clue what they're doing and even if they do accidentally introduce the most obscure bugs and unpredictable behavior.

Symfony does not do magic. `autowire` edges on magic, and is thus even in official docs advised not to use in production applications. `autoconfigure` is the milder variant, explicitly subscribing you to sane suggested default configs. Because you explicitly chose to.

**Comment by stof** [comment]
and having it on by default would impact all existing code, and so would be a BC break.

**Comment by TomasVotruba** [comment]
@curry684 Well, tagging is magic, no doubt about that. This only adds magic over magic.

But I rather look from a practical point of view. A use case,  **when you don't want this feature on by default?** Could you give me some?

**Comment by curry684** [comment]
That's a very easy one actually - when following [standard procedure by putting Console commands in the `Command` namespace of a bundle](http://symfony.com/doc/current/console/commands_as_services.html). You don't need to tag them there.

But for example for Security Voters I may have an application which defines 20 voters of which, depending on deployment type or the availability of external integrations, only 15, 17 or 19 are actually going to be enabled. Not tagging them takes them out of use.

For Event Subscribers the same goes - I may opt to disable a few low level logging event subscribers in certain deployments for performance reasons, and tag them based on env variables in my bundle compilation.

It's not our job to tell the developers what they want, we're just providing the enabling tools.

**Comment by TomasVotruba** [comment]
> That's a very easy one actually - when following standard procedure by putting Console commands in the Command namespace of a bundle. You don't need to tag them there.

That's just miss leading of [another magic](https://php-and-symfony.matthiasnoback.nl/2013/10/symfony2-console-commands-as-services-why/). And You do, if you use DI.

> But for example for Security Voters I may have an application which defines 20 voters of which, depending on deployment type or the availability of external integrations, only 15, 17 or 19 are actually going to be enabled. Not tagging them takes them out of use.

Why would you register a service you don't use?

>  I may opt to...

"In theory everything is true." Do you actually already do that?

This sound like architecture smell in your application. To much coupling and breaking SRP.

> It's not our job to tell the developers what they want...

That's what frameworks do actually. If you get hammer, everything seem like a nail.

Still missing some saint and practical examples.

**Comment by curry684** [comment]
> Why would you register a service you don't use?

Why shouldn't I be able to register a voter I wish to invoke manually and conditionally? The case is also perfectly sane and practical in a bundle to have an extension only tag a number of specific voters if configuration says a specific external authorization component is available.

> That's what frameworks do actually. If you get hammer, everything seem like a nail. 

No, this is definitely not true of frameworks like .NET and Symfony. They provide you with a huge toolbox of things, and then allow you to use *not a single one of them*, or just the hammer, or just the screwdriver, or both. That's why they're classified as **unopinionated**, you the developer provide the opinions by opting into functionality you want to be there. Batteries are included but optional and replaceable.

Laravel on the other hand is a clear example of an **opinionated** framework. Their opinions make for fast development of predictable and rigidly structured projects. But if you don't agree with a significant number of their opinions, you shouldn't be developing in Laravel. You cannot realistically replace the battery without breaking everything else.

**Comment by TomasVotruba** [comment]
> Why shouldn't I be able to register a voter I wish to invoke manually and conditionally? The case is also perfectly sane and practical in a bundle to have an extension only tag a number of specific voters if configuration says a specific external authorization component is available.

Could you paste a code from your app where you do that? Still seems like overcomplicating simple things easily solvable by good service design.

> No, this is definitely not true of frameworks like .NET and Symfony.

This is so common claim among Symfony developers I've noticed in last few years. Many people that use Symfony for years are very biased. I was the same when using one framework for years. It's common confirmation bias. Symfony forces me to use lot of over-complicated code just to register simple service and load it into another.

> Laravel on the other hand is a clear example of an opinionated framework.

As many Symfony contribution are made ~5 **very similar people**, it's very opinionated framework.
One things a like about Laravel is that it gives you much greater freedom and doesn't vendor lock as much as Symfony. That's where Symfony could learn a lot.

Like in this feature. Symfony is getting very complicated for simple service usage with not much advantage of end users.

**If there would be one optional place to turn this on, it would make sense, but this just makes my polute my configs.**

**Comment by TomasVotruba** [comment]
I care about this because these features I present during my lectures and mentoring. And people give me WTF and I have no useful reasons for it. I don't want to teach people features just because they exists. I want to teach them the best known approach on the market.

**I feel like Symfony is to be designed more and more for people that already use it for years, and closing doors more and more for new ones**. That's probably why Laravel is doing so great.

**Comment by theofidry** [comment]
@TomasVotruba not sure from where you get that feeling. As @stof said it's simply not possible to enable it by default as it would be a BC. And I think on the contrary Symfony is changing a lot to be easier to use and more friendly to people coming in. It strives at finding a good balance between providing the strictness for the anti-magic people and the magic all the way people. And from what I've seen in the 3.x, it's on good track for it.

**Comment by GuilhemN** [comment]
Note that this feature will be enabled by default by Flex (see https://github.com/symfony/recipes/blob/master/symfony/framework-bundle/3.3/etc/packages/app.yaml#L4), this makes your config file much simpler than before while still being explicit, I think it's the best compromise to satisfy everyone.

**Comment by TomasVotruba** [comment]
> @TomasVotruba not sure from where you get that feeling. 

Mostly from recently added "config programming" features in 3.3, getter autowiring, per config autotowiring etc.

> As @stof said it's simply not possible to enable it by default as it would be a BC.

Where do get feeling I want a BC break? I want an **option** to enable it for all services. It can be off by default, of course - it should.

Maybe use case would be better.
 
In previous project we had over 30 modules. When I use these features, I will end up with:

```yaml
# src/*Module/config/services.yml
services:
    _defaults:
        autowire: true
        autoconfigure: true # this adds tags actually

   # services
   // ...
```

30x times repeated everywhere. DRY? SRP? SOLID? I don't think so.
I also have to teach every person in the project to use this "perex" in every config that is related to adding services in any way.

I prefer writing only what is necessary.

```yaml
# src/*Module/config/services.yml
services:
   # services
   // ...
```

Somebody will definitely forget (we all  know these bugs) to add this to another `services.yml` and will be surprise to see one Voter to be registered and other now.


Why creating a new way to confuse people? I really don't see a point to make Symfony this playground.

This feature is already covered in code of FrameworkBundle anyway:

```yaml
# app/config/config.yml
framework:
    autoconfigure: true # add tags to services
```

Saved possibly millions of lines and thousands of bug reports in all applications in the worlds. I'm sure, because I've been consulting and mentoring various Symfony/Nette/proprietary projects and this kind of stuffs takes the most time. Not to setup, but to debug eventually, when team grows and complexity rises.

I think this is difficult to spot, when working on 1 framework and 1 company, yet when all tests are passing, right. I would noticed it before either. But from my experience and feedback from programmers from my consultancy work, these are double-edged feature more giving now, but more taking later.

**Comment by stof** [comment]
A global option is a pain in the ass, because it also impacts all code which is not yours (symfony itself, external bundles, etc...). Having a different behavior for your service configuration depending on some project level configuration would make it a real pain for all bundle maintainers (as they would have to overwrite all config to repeat the Symfony default to ensure it is indeed the default being applied, which then becomes a pain to support several Symfony versions as the list of config settings you need to define is different for each version).
This is why we all hate PHP features which can change their behavior based on a INI setting (and why several such INI settings have been deprecated in recent PHP versions)

**Comment by TomasVotruba** [comment]
> A global option is a pain in the ass...

Can be limited to app of course, that's just implementation detail.

I just autowiring for over 6 years in Nette and Symfony apps with no problems.

**Comment by theofidry** [comment]
@TomasVotruba I don't think hijacking this PR is the right place to debate about that :)

That said, the problem I see in your comments come from:

- You are using autowiring everywhere
- You are using an extremely large amount of service files

For the first point it's your choice, but then I don't understand the later: you can simply register all the source code of your application as services which would simulate what you have with a DIC which has autowiring by default: you have to register a very limited amount of services and you don't have to re-configure everything.

But as I said, it's not the best place to discuss about that. Maybe this would be more appropriate on Slack (https://symfony.com/slack-invite) or a dedicated issue :)

**Comment by TomasVotruba** [comment]
@theofidry My intention is not to hijack PR, why do you think so?

I want to point out bad design of these features and underline them with my experience and feedback from Symfony users around me. My goal is to improve usability of Symfony, so people are not forced to use more developer friendly frameworks.

> You are using autowiring everywhere.

That's a design, not a problem.

> You are using an extremely large amount of service files.

What do you mean? Do prefer having monolithic design with huge coupled files?

Feel free to open issue, I'm not much into private Slack talks about PRs.

**Comment by theofidry** [comment]
>My intention is not to hijack PR, why do you think so?
>I want to point out bad design of these features and underline them with my experience and feedback from Symfony users around me. My goal is to improve usability of Symfony, so people are not forced to use more developer friendly frameworks.

Hence a dedicated issue as it's much broader than the specific feature of this PR which has already been merged. 

>That's a design, not a problem.

Bad phrasing, soz.

>What do you mean? Do prefer having monolithic design with huge coupled files?

Ha, that's where I'm asking you what you mean. It's quite common, auto-wiring or not, to have 1 `services.yml` file. Which I agree can end up quite big which is why I prefer to break it up. But if you are using autowiring everywhere, its size should significantly decrease.

But maybe having more concrete examples here would help :)

**Comment by TomasVotruba** [comment]
> Hence a dedicated issue as it's much broader than the specific feature of this PR which has already been merged.

I've learned about this from a blog post, thus writing now.

>  It's quite common, auto-wiring or not, to have 1 services.yml file.

It seems that the false axiom, that you should reconsider to validate. 

I've seen single `services.yml` only in small, coupled or legacy applications. Not a profitable way to go in long term team size scaling.

To give you a scale: I work with applications old 3-10 years having 500-1500 services (on average).

Never the less, single `services.yml` **promotes** coupling-first, even when not needed to. I mean it's good for Symfony, because it makes very expensive to switch to another framework - and I'm aware that's strategy of almost every framework - but bad for users, because they have to learn decoupling the hard way.

> But maybe having more concrete examples here would help :)

What other examples do you need? [The one above](https://github.com/symfony/symfony/pull/22234#issuecomment-297987265) describes the issue I talk about. I can complete services if you need :D 

> Hey, you're the one who wants to voice your concerns.

I'm doing so here and now.

**Comment by theofidry** [comment]
>To give you a scale: I work with applications old 3-10 years having 500-1500 services (on average).

Yeah, and I've already seen that in not more than 3 service files. I don't recommend to do it though...

>Never the less, single services.yml promotes coupling-first

I don't understand how your service declaration promotes coupling. If you want to switch of framework you'll have to rewrite your service wiring (auto-wiring involved or not) and that regardless of the number of services/files.

>>But maybe having more concrete examples here would help :)

>What other examples do you need? The one above describes the issue I talk about. I can complete services if you need :D

```yaml
services:
    App\:
        resource: '%kernel.root_dir%/../src'
        autowire: true
```

Or something along those lines, then you only need to declare the a few services which require dedicated config. With that, I don't see how you run into the issue of having to repeat those file defaults config 30 times over.

**Comment by TomasVotruba** [comment]
> If you want to switch of framework you'll have to rewrite your service wiring (auto-wiring involved or not) and that regardless of the number of services/files.

Not really. Well written apps are easy to switch, since you only you routing and kernel logic.
Everything else can and should be decoupled. It's difficult to think that way, took me years, more over when frameworks are coupling almost everything by default.

> I don't understand how your service declaration promotes coupling.

Let's say you have one e-commerce application that contains ProductBundle (or module, extension whatever). It uses `app/config/services.yml` as you recommend. In time, you decouple CategoryBundle. Then TagBundle and many more.

Then you want to use CategoryBundle in other project, dealing with online content. If you write your code  with modularity in mind, it only takes adding few interfaces. 

But what if you have one huge `app/config/services.yml`? It would take creating new Bundle, Extension, CompilerPasses etc., so many people just copy-paste code. A code, that is 80 % ready to reuse.

### Everything over adding a Bundle to AppKernel is redundant

I don't want to and take care of config. I want to plugin and code. When you have decouple application and 2-3 projects are going to share few logic, or microservices splitting, one single `app/config/services.yml` is killer for any change.

I've also seen few cases, when somebody added bundle, but didn't register `services.yml` or `routing.yml`. And application run very well. Well... apart few routes and services :D Pattern issue, not programmer.

### Why not use config for everything?

Setting up a module is usually more complicated than adding services, like you mention.

Usually, you need to join them and add some parameters.

**Again, it's about reusability of growing app that need long-term support**. If there would be no bundles, extensions and no applications that might share a code, using one `app/config/config.yml` would be great way to go. But, for such small and dummy app I would use Laravel or Symfony\Console instead.

**Comment by TomasVotruba** [comment]
I'm curious and could not find it, where was `resource` service autodiscovery added? PR or Post would be great. Thanks

```yaml
services:
    App\:
        resource: '%kernel.root_dir%/../src'
```

**Pull request closed without merging**
